### PR TITLE
Implement Sandbox notifications processor and publisher

### PR DESCRIPTION
**Pull request by Future-Outlier**
# TL;DR
Implement notifications processor and publisher for Sandbox.

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [ ] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
Add new **SandboxProcessor** and **SandboxPublisher** that implement the Processor interface.

I add config in yaml file like this
```
notifications:
  type: sandbox
  emailer:
    emailServerConfig:
      apiKeyEnvVar: "SENDGRID_API_KEY"
    subject: "Notice: Execution \"{{ name }}\" has {{ phase }} in \"{{ domain }}\"."
    sender: "example@gmail.com"  # Consider using a generic email instead of personal
    body: >
      Execution \"{{ name }}\" has {{ phase }} in \"{{ domain }}\". View details at
      <a href=\http://example.com/projects/{{ project }}/domains/{{ domain }}/executions/{{ name }}>
      http://example.com/projects/{{ project }}/domains/{{ domain }}/executions/{{ name }}</a>. {{ error }}
```
screenshot
image

## Tracking Issue
_Remove the '*fixes*' keyword if there will be multiple PRs to fix the linked issue_

fixes https://github.com/flyteorg/flyte/issues/<number>

## Follow-up issue
_NA_
OR
_https://github.com/flyteorg/flyte/issues/<number>_

### CODE REVIEW

File: pkg/async/notifications/implementations/sandbox_processor.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+		}
+	}
+}
+
+func (p *SandboxProcessor) StopProcessing() error {
+	logger.Debug(context.Background(), "call to sandbox stop processing.")
+	return nil
+}
+
+func NewSandboxProcessor(emailer interfaces.Emailer) interfaces.Processor {
```
Comment (pingsutw) [inline pkg/async/notifications/implementations/sandbox_processor.go:55]: Could we add a test for the sandbox processor that uses the default noop emailer?

File: pkg/async/notifications/factory.go [hunk 2]
```diff
@@ -66,6 +66,8 @@ func GetEmailer(config runtimeInterfaces.NotificationsConfig, scope promutils.Sc
 			scope,
 			sesClient,
 		)
+	case common.Sandbox:
+		return implementations.NewSendGridEmailer(config, scope)
```
Comment (pingsutw) [inline pkg/async/notifications/factory.go:70]: IIUC, we'll create SendGridEmailer at line 50, so we don't need these two lines here.

File: pkg/async/notifications/implementations/sandbox_processor.go (hunk 1 above)
Comment (Future-Outlier) [reply pkg/async/notifications/implementations/sandbox_processor.go:55]: Yes, I will try it

File: pkg/async/notifications/factory.go (hunk 2 above)
Comment (Future-Outlier) [reply pkg/async/notifications/factory.go:70]: Yes, but you need to turn your config to 
```
emailServerConfig:
      serviceName: sendgrid
      apiKeyEnvVar: "SENDGRID_API_KEY"
```
with one more key in serviceName

**Comment by Future-Outlier** [comment]
@pingsutw  I've updated my test, please check it.
Also to use the mock emailer int test function, I've added a "return nil" in my sandbox_processor.go.
Thanks a lot.

**Comment by Future-Outlier** [comment]
I am not pretty sure whether use noop emailer or mock emailer, please contact me if it is needed to change the code.
Thanks a lot.

**Comment by Future-Outlier** [comment]
BTW, I just copy the master branch's go.mod and go.sum to my branch currently and run "go mod tidy"
I found that it is not needed to update the go.mod and go.sum.
Thanks a lot

File: pkg/async/notifications/implementations/sandbox_publisher_test.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestSandboxPublisher_Publish(t *testing.T) {
+	publisher := NewSandboxPublisher()
```
Comment (pingsutw) [inline pkg/async/notifications/implementations/sandbox_publisher_test.go:12]: Can we use NewNotificationsPublisher to create SandboxPublisher here?

File: pkg/async/notifications/implementations/sandbox_processor_test.go [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+	assert.Equal(t, emailMessage.Body, testEmail.Body)
+	assert.Equal(t, emailMessage.RecipientsEmail, testEmail.RecipientsEmail)
+	assert.Equal(t, emailMessage.SubjectLine, testEmail.SubjectLine)
+	assert.Equal(t, emailMessage.SenderEmail, testEmail.SenderEmail)
+}
+
+func TestSandboxProcessor_StartProcessing(t *testing.T) {
+
+	testSandboxProcessor := NewSandboxProcessor(&mockSandboxEmailer)
```
Comment (pingsutw) [inline pkg/async/notifications/implementations/sandbox_processor_test.go:28]: Same, Can we use NewNotificationsProcessor to create SandboxProcessor here?

**Review by pingsutw — APPROVED** [review]

File: pkg/async/notifications/implementations/sandbox_publisher_test.go (hunk 3 above)
Comment (Future-Outlier) [reply pkg/async/notifications/implementations/sandbox_publisher_test.go:12]: @pingsutw 
I've tried it, but the import cycle is not allowed in go test.

```
imports github.com/flyteorg/flyteadmin/pkg/async/notifications: import cycle not allowed in test
```
I've wrote code like this
```
var scope = promutils.NewScope("test_sandbox_processor")
notificationsConfig := runtimeInterfaces.NotificationsConfig{
Type: "sandbox",
}
testSandboxProcessor := notifications.NewNotificationsProcessor(notificationsConfig, scope)
```

And the dependencies I will list at here
```
package github.com/flyteorg/flyteadmin/pkg/async/notifications/implementations
imports github.com/flyteorg/flyteadmin/pkg/async/notifications: import cycle not allowed in test
```

If test the NewNotificationsPublisher  is necessary, I can try it in another package, but I am not sure whether this is what you want, thanks a lot

Comment (pingsutw) [reply pkg/async/notifications/implementations/sandbox_publisher_test.go:12]: I see, ok, that's fine

File: pkg/async/notifications/factory_test.go [hunk 5]
```diff
@@ -23,3 +31,13 @@ func TestGetEmailer(t *testing.T) {
 	// shouldn't reach here
 	t.Errorf("did not panic")
 }
+
+func TestNewNotificationsProcessor(t *testing.T) {
+	testSandboxProcessor := NewNotificationsProcessor(notificationsConfig, scope)
```
Comment (pingsutw) [inline pkg/async/notifications/factory_test.go:36]: Nit: could you call StartProcessing and StopProcessing here?

Comment (Future-Outlier) [reply pkg/async/notifications/factory_test.go:36]: No problem

**Review by pingsutw — APPROVED** [review]
LGTM, Thanks!

**Comment by Future-Outlier** [comment]
> LGTM, Thanks!

Thanks a lot for guiding me ^^

**Review by wild-endeavor — APPROVED** [review]

File: pkg/async/notifications/implementations/sandbox_publisher.go [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+import (
+	"context"
+
+	"github.com/flyteorg/flytestdlib/logger"
+	"github.com/golang/protobuf/proto"
+)
+
+type SandboxPublisher struct{}
+
+var msgChan = make(chan []byte)
```
Comment (katrogan) [inline pkg/async/notifications/implementations/sandbox_publisher.go:12]: can we instantiate this in the `NewSandboxPublisher()` constructor and make it a member variable? Should make things easier to unit test too!

Comment (katrogan) [reply pkg/async/notifications/implementations/sandbox_publisher.go:12]: Alternatively we can pass it to NewSandboxPublisher() and NewSandboxProcessor()

Comment (Future-Outlier) [reply pkg/async/notifications/implementations/sandbox_publisher.go:12]: I think I can implement the alternative solution.
But how to pass the msgChan to first solution ?
Thanks a lot.

Comment (katrogan) [reply pkg/async/notifications/implementations/sandbox_publisher.go]: Sure, you could do 

```
// Receiver only channel
func NewSandboxProcessor(msgChan <-chan []byte)  

// Sender only channel
and NewSandboxPublisher(msgChan chan<- []byte)
```

And continue to initialize e chan it outside of the processor and publisher

File: pkg/async/notifications/implementations/sandbox_processor.go [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+		case msg := <-p.msgChan:
+			err := proto.Unmarshal(msg, &emailMessage)
+			if err != nil {
+				logger.Errorf(context.Background(), "error with unmarshalling message [%v]", err)
+				return err
+			}
+
+			err = p.email.SendEmail(context.Background(), emailMessage)
+			if err != nil {
+				logger.Errorf(context.Background(), "error with sendemail message [%v] ", err)
```
Comment (katrogan) [inline pkg/async/notifications/implementations/sandbox_processor.go:42]: can we log the failed email message here too?

File: pkg/async/notifications/implementations/sandbox_processor_test.go [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+	"github.com/flyteorg/flyteadmin/pkg/async/notifications/mocks"
+	"github.com/flyteorg/flyteidl/gen/pb-go/flyteidl/admin"
+	"github.com/golang/protobuf/proto"
+	"github.com/stretchr/testify/assert"
+)
+
+var mockSandboxEmailer mocks.MockEmailer
+
+func TestSandboxProcessor_UnmarshalMessage(t *testing.T) {
+	var emailMessage admin.EmailMessage
```
Comment (katrogan) [inline pkg/async/notifications/implementations/sandbox_processor_test.go:16]: What's the purpose of this test? Are we just verifying we can unmarshal a proto? Considering that test code already initializes this msg var I'm not sure how valuable this unit test is.

File: pkg/async/notifications/implementations/sandbox_processor.go (hunk 7 above)
Comment (Future-Outlier) [reply pkg/async/notifications/implementations/sandbox_processor.go:42]: Yes I will

File: pkg/async/notifications/implementations/sandbox_publisher.go (hunk 6 above)
Comment (Future-Outlier) [reply pkg/async/notifications/implementations/sandbox_publisher.go]: Can I clarify what "continue to initialize msgChan" means?
I thought that I just have to initialize the msgChan 1 time only.

File: pkg/async/notifications/implementations/sandbox_processor_test.go (hunk 8 above)
Comment (Future-Outlier) [reply pkg/async/notifications/implementations/sandbox_processor_test.go:16]: Yes, you are right, I can consider deleting it.

**Review by pingsutw — DISMISSED** [review]

**Review by katrogan — APPROVED** [review]
Looks great, I think the code cov annotations have suggestions for some additional test cases we could use to bump the code coverage on this PR and so that we can test more scenarios

**Comment by Future-Outlier** [comment]
> Looks great. This is failing the code coverage check, but I think the code cov annotations have suggestions for some additional test cases we could use to bump the code coverage on this PR and so that we can test more scenarios

Thanks for the suggestions, I've learned a lot!

**Review by pingsutw — APPROVED** [review]

**Review by katrogan — APPROVED** [review]
Thank you for adding all the unit tests!
